.github/workflows/test.yml

```diff
@@ -414,3 +436,13 @@ jobs:
       with:
         name: runtests-${{ runner.os }}-py${{ matrix.python-version }}-Salt${{ matrix.salt-version }}.log
         path: artifacts/runtests-*.log
+
+    - name: Publish distribution 📦 to Test PyPI
+      if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
+      uses: pypa/gh-action-pypi-publish@release/v1
+      with:
+        repository_url: https://test.pypi.org/legacy/
+        print-hash: true
+        skip-existing: true
+        verbose: true
+        verify-metadata: true
```

Comment:
```suggestion
  Build:
    runs-on: ubuntu-latest
    environment: testing
    permissions:
      id-token: write
    needs:
      - macOS
      - Windows
      - Linux
      - Docs
    steps:
    - uses: actions/checkout@v4
      with:
        fetch-depth: 0
    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: "3.10"
    - name: Install Nox
      run: |
        python -m pip install nox
    - name: Build a binary wheel and a source tarball
      run: |
        nox -e build

    - name: Publish distribution 📦 to Test PyPI
      if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
      uses: pypa/gh-action-pypi-publish@release/v1
      with:
        repository_url: https://test.pypi.org/legacy/
        print-hash: true
        skip-existing: true
        verbose: true
        verify-metadata: true
```

- There should be a separate step doing the build/publish after all other tests have passed, ensuring this only runs once and for one particular version of Python.
- THOUGH: Keep in mind, this is assuming `nox -e build` is how the package is built. You can update my suggestion based on whether a different command should be use.
- My suggestion is also assuming `testing` is the name of the env on GitHub, so can modify the `environment: testing` suggestion if needed
- I'm also using `python-version: "3.10"` as a way of ensuring the uploaded build is done on the version Salt uses in 3006+

ALSO: I think maybe we should evaluate whether it would make sense to do something like a separate workflow that triggers a publish to `test` PYPI whenever a `push` to `main` is done, if we care to have it separate from acting on PRs?

Reply:
okay will look at doing that tomorrow

Reply:
Implemented your suggestion, and researching using ```testing``` but worried about tools other people like NSX in India are using and if doing that would break them.

Do you have any examples of seperate workflows for publish whenever a ```push``` to ``` main``` occurs ?

Reply:
I think it would look something like this:

```yaml
name: Test PyPi Publish

on:
  push:
    branches:
      - main

jobs:
  Publish:
    runs-on: ubuntu-latest
    environment: testing # Or whatever you name your github environment
    permissions:
      id-token: write
    steps:
    - uses: actions/checkout@v4
      with:
        fetch-depth: 0

    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: "3.10"

    - name: Install Nox
      run: |
        python -m pip install nox
    - name: Build a binary wheel and a source tarball
      run: |
        nox -e build

    - name: Publish distribution 📦 to Test PyPI
      if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
      uses: pypa/gh-action-pypi-publish@release/v1
      with:
        repository_url: https://test.pypi.org/legacy/
        print-hash: true
        skip-existing: true
        verbose: true
        verify-metadata: true
```

Up to us on if we'd want the `if: github.event_name...` conditional check, since if it is present, it would mean it would only trigger the publishing step to Test PyPi if it was a tag. If we keep that check, it would mean we could use a tag that is an `rc` type release version, since that would show whether the publishing to Test PyPi works, helping ensure that a GitHub Release trigger for prod PyPi would have a higher likelihood of known success after.

Reply:
Done